src/pegen/python_generator.py

```diff
@@ -299,12 +316,21 @@ def visit_Rule(self, node: Rule) -> None:
         if node.name.endswith("without_invalid"):
             self.cleanup_statements.pop()
 
-    def visit_NamedItem(self, node: NamedItem) -> None:
+    def visit_NamedItem(
+        self, node: NamedItem, used: Optional[Set[str]], unreachable: bool
+    ) -> None:
         name, call = self.callmakervisitor.visit(node.item)
-        if node.name:
+        if unreachable:
+            name = None
+        elif node.name:
             name = node.name
+
+        if used is not None and name not in used:
+            name = None
+
         if not name:
-            self.print(call)
+            # Parentheses are needed because the trailing comma may appear :>
+            self.print(f"({call})")
```

Comment:
Could you just point me to a concrete example of this issue ?

Reply:
Woulld love to!
You cannot find it in the metaparser, but in the python parser.

Reply:
For example:
```python
@memoize
    def file(self) -> Optional[ast . Module]:
        # file: statements? $
        mark = self._mark()
        if (
            (a := self.statements(),)
            and
            (self.expect('ENDMARKER'))
        ):
            return ast . Module ( body = a or [] , type_ignores = [] );
        self._reset(mark)
        return None;
```

Reply:
As it seems all of the rules like `X?` will result in this

Reply:
And `[X]` will also result in this

Reply:
It seems logical, optional rule cannot be incorrect as if there's nothing found it's still fine!

Reply:
I need to look at how things work currently and why this is now needed.

Reply:
So I checked and in the case of the file rule the parentheses are already there for the case of statements so I believe we do not need this change or am I missing something ?

Reply:
parentheses are there because the name is used so
`self.print(f"({name} := {call})")` is responsible
But if we don't use the name we still need to have parentheses to handle the comma
`(X := y,) == (y,)`

Reply:
Generate a python parser with this branch and search for `,)` you'll see the examples of both of those cases

Reply:
Now I get it. I just wished we could do it in such a way we do not generate extra parenthesis but I guess black can clean up for us. Thanks for bearing with me.

Reply:
You're welcome!
Why do you not like generating parentheses in there?

Reply:
The ones around the endmarker in the example above look funny to me but it is not a big deal.